Fix buffer leak in `HttpClientStreamTest.testUploadWithConcurrentServerCloseClosesStream()`

lorban opened the pull request:
Depends on https://github.com/jetty/jetty.project/pull/10432 - so this PR is a branch of that other PR.

See #10226

lorban committed:
#10226 force completion of the handle's callback to fix a buffer leak

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
This requires more investigations as stopping the connector should release any pending buffer not yet handed over to user code.

lorban committed:
#10226 for releasing the buffer and failing the stream in Connection.onClose()

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10226 only fail the stream when a cause is given

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10226 only fail the stream when a cause is given

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10226 release retainable buffer atomically

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10226 enable leak tracking

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@sbordet nudge

sbordet commented on jetty-core/jetty-server/src/test/java/org/eclipse/jetty/server/AbstractHttpTest.java:
```diff
@@ -57,8 +60,16 @@ public void setUp() throws Exception
     {
-        server.stop();
-        stacklessChannelLogging.close();
+        try
+        {
+            Set<ArrayByteBufferPool.Tracking.Buffer> serverLeaks = bufferPool.getLeaks();
+            assertEquals(0, serverLeaks.size(), bufferPool.dumpLeaks());
+        }
+        finally
+        {
+            server.stop();
+            stacklessChannelLogging.close();
+        }
```
Can you re-arrange to avoid the `finally`...
getLeaks()
stop()
close()
assert()

lorban replied:
Done.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -108,7 +108,7 @@ public class HttpConnection extends AbstractConnection implements Runnable, Writ
     private final AtomicBoolean _handling = new AtomicBoolean(false);
     private final HttpFields.Mutable _headerBuilder = HttpFields.build();
-    private volatile RetainableByteBuffer _retainableByteBuffer;
+    private final AtomicReference<RetainableByteBuffer> _retainableByteBuffer = new AtomicReference<>();
```
Why was it necessary to use an atomic here?
As this field is "hot", so avoiding an indirection would be good.

However, the changes to atomic revealed that the code before was not really paying attention to the fact that the field was `volatile`, so the code is now better, but with an indirection.
Can we have better without indirection?

lorban replied:
I agree that `_retainableByteBuffer` should not need to be thread-safe, and I cannot remember exactly where I found a race condition that made me write this change.

Let me investigate, at the very least the reasoning should be documented but ideally, this field should not require any thread-safety.

lorban committed:
#10226 handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
#10226 tentative: revert buffer's thread safety

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
On hold until #10277 is sorted.

lorban commented:
The problem is that stopping the connector is notifying the HttpConnection via onClose, but that notification isn't reported to the channel state.

The tentative fix of releasing the buffer in onClose and ensuring the buffer only gets released once by making it atomic isn't right as HttpConnection's buffer shouldn't need any thread safety.

joakime commented:
@lorban I think this PR needs to be updated from `jetty-12.0.x`

lorban committed:
- fix HttpClientIdleTimeoutTest to wait for server's idle timeout before checking for leaks
- improve HttpClientIdleTimeoutTest by making it upload some content
- fix FCGI server leak caused by idle timeout
- fix H3 server leak caused by idle timeout

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime commented:
@lorban this PR needs a merge from `jetty-12.0.x` HEAD.

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
Merge branch 'fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
fix missing acquire

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
- fix buffer leak in H1, FCGI when connector is stopped with pending requests
 - re-enable leak tracking on fixed tests

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/fix/jetty-12-10226-fcgi-HttpClientIdleTimeoutTest-leaks' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
add investigation notes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
fix compilation

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
remove unneeded change

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert unneeded test changes + fix tracking pool ctors

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert unneeded test changes + fix tracking pool ctors

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
revert unneeded test changes + fix tracking pool ctors

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix atomic ref access

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
keep H1 leak tracking disabled + explain why

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban commented:
@sbordet nudge

lorban committed:
remove H1 changes

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
better fix?

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

sbordet commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -407,4 +408,12 @@ public void onFailure(int request, Throwable failure)
+    public void close()
+    {
+        if (stream != null)
+            stream.getHttpChannel().onFailure(new EofException());
```
`onFailure()` returns a `Runnable` that must be run.

lorban replied:
Done.

sbordet commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -610,6 +610,13 @@ public boolean onIdleExpired(TimeoutException timeout)
+    @Override
+    public void close()
+    {
+        _httpChannel.onFailure(new EofException());
```
Ditto.

lorban replied:
Done.

lorban committed:
Merge remote-tracking branch 'origin/jetty-12.0.x' into fix/jetty-12-10226-testUploadWithConcurrentServerCloseClosesStream

lorban committed:
execute runnable + add test asserting the executable did run

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
fix test

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

joakime reviewed (dismissed):
Anything holding back this from being merged?

lorban commented:
@joakime not that I know. A final review pass from @sbordet would be preferable but I think I addressed all his concerns.

sbordet commented on jetty-core/jetty-tests/jetty-test-client-transports/src/test/java/org/eclipse/jetty/test/client/transport/HttpClientStreamTest.java:
```diff
@@ -1065,6 +1059,99 @@ public void close() throws IOException
+    @MethodSource("transports")
+    public void testUploadWithPendingReadConcurrentServerCloseClosesStream(Transport transport) throws Exception
+    {
+        CountDownLatch serverDemandLatch = new CountDownLatch(1);
```
What's this for?
Should you not count it down in the server-side demand callback?

lorban replied:
Yes, this latch should be counted down after `lastChunk` has been set and awaited before reading it.

lorban committed:
handle review comments

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpConnection.java:
```diff
@@ -610,6 +610,15 @@ public boolean onIdleExpired(TimeoutException timeout)
+    @Override
+    public void close()
+    {
+        Runnable task = _httpChannel.onFailure(new EofException());
```
I don't like creating an `EofException` for every normal close.   
If you really need this, then create a `_httpChannel.onClose()` method that grabs the state lock, checks the state and only creates the exception if it is actually needed.

lorban replied:
I added an extra check so the channel is only failed when `_httpChannel.getRequest()` is not null, such as the channel is only failed (and the exception created) only if there is a pending request, which is not the case for a normal close.

That makes the implementation a bit more elegant and takes care of that unneeded exception allocation in the normal case.

gregw commented on jetty-core/jetty-fcgi/jetty-fcgi-server/src/main/java/org/eclipse/jetty/fcgi/server/internal/ServerFCGIConnection.java:
```diff
@@ -407,4 +408,16 @@ public void onFailure(int request, Throwable failure)
+    {
+        if (stream != null)
+        {
+            Runnable task = stream.getHttpChannel().onFailure(new EofException());
```
See comment on HttpConnection... but I care a little less about FCGI

lorban replied:
I applied the same change I made in `HttpConnection`.

lorban committed:
only fail the channel during close if there is a pending request

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
introduce HttpChannel.onClose

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

lorban committed:
minor test improvement

Signed-off-by: Ludovic Orban <lorban@bitronix.be>

gregw commented on jetty-core/jetty-server/src/main/java/org/eclipse/jetty/server/internal/HttpChannelState.java:
```diff
@@ -463,6 +464,21 @@ public Runnable onFailure(Throwable x)
+                LOG.debug("onClose {} stream={}", this, _stream);
+
+            // If the channel doesn't have a stream, then no action is needed.
+            if (_stream == null)
```
would `_request == null` be a better test.    Do we want an EofException if a connection is opened then closed without a request?

lorban replied:
I used the `_stream == null` test for consistency with `onFailure()` that tests for that too, but in practice it's identical to `_request == null` as both are reset to null at the same place in `recycle()`.

And I do believe it is correct to fail with a `EofException` when a connection is opened then immediately closed without sending a byte.

@sbordet Do you agree?

gregw reviewed (approved):
Conditional approval... with one question
